### PR TITLE
fix(service-worker): ensure initialised in browser only

### LINKED ISSUE #20360: ServiceWorkerModule.register is not Universal friendly.
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

When having setup the ServiceWorkerModule in your AppModule it crashes when the server is trying to render the application because of the `navigator` object not being defined in the server.

## Expected behavior

No crash.

## Minimal reproduction of the problem with instructions

`ReferenceError: navigator is not defined
    at Array.initializer (C:\Users\Stan\Documents\bravobit-web\functions\node_modules\@angular\service-worker\bundles\service-worker.umd.js:373:34)
    at ApplicationInitStatus.runInitializers (C:\Users\Stan\Documents\bravobit-web\functions\node_modules\@angular\core\bundles\core.umd.js:3490:67)
    at C:\Users\Stan\Documents\bravobit-web\functions\node_modules\@angular\core\bundles\core.umd.js:5418:28
    at _callAndReportToErrorHandler (C:\Users\Stan\Documents\bravobit-web\functions\node_modules\@angular\core\bundles\core.umd.js:5597:39)
    at C:\Users\Stan\Documents\bravobit-web\functions\node_modules\@angular\core\bundles\core.umd.js:5416:20
    at ZoneDelegate.invoke (C:\Users\Stan\Documents\bravobit-web\functions\node_modules\zone.js\dist\zone-node.js:392:26)
    at Object.onInvoke (C:\Users\Stan\Documents\bravobit-web\functions\node_modules\@angular\core\bundles\core.umd.js:4668:33)
    at ZoneDelegate.invoke (C:\Users\Stan\Documents\bravobit-web\functions\node_modules\zone.js\dist\zone-node.js:391:32)
    at Zone.run (C:\Users\Stan\Documents\bravobit-web\functions\node_modules\zone.js\dist\zone-node.js:142:43)
    at NgZone.run (C:\Users\Stan\Documents\bravobit-web\functions\node_modules\@angular\core\bundles\core.umd.js:4485:69)
`

## What is the motivation / use case for changing the behavior?

The server should not crash because of a undefined property.

## Environment

<pre><code>
Angular version: 5.0.1

Browser:
- [x] Chrome (desktop) version XX
- [x] Chrome (Android) version XX
- [x] Chrome (iOS) version XX
- [x] Firefox version XX
- [x] Safari (desktop) version XX
- [x] Safari (iOS) version XX
- [x] IE version XX
- [x] Edge version XX
 
For Tooling issues:
- Node version: 8.8.1 
- Platform: Windows

Others:

</code></pre>

### PR DESCRIPTION
**Pull request by Toxicable**
closes #20360 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
```
[x] Bugfix
```

## What is the current behavior?
Using ServiceWorkerModule on the server would cause an error due to `navigator` being undefined

Issue Number: 20360

## What is the new behavior?
Only runs the initialisation function if running on browser

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other
Not really sure how to test this since it would require a platform-server integration test and I cant find any existing ones for the SW

### CODE REVIEW

File: packages/service-worker/src/module.ts [hunk 1]
```diff
@@ -20,10 +21,11 @@ export const SCRIPT = new InjectionToken<string>('NGSW_REGISTER_SCRIPT');
 export const OPTS = new InjectionToken<Object>('NGSW_REGISTER_OPTIONS');
 
 export function ngswAppInitializer(
-    injector: Injector, script: string, options: RegistrationOptions): Function {
+    injector: Injector, script: string, options: RegistrationOptions,
+    platformId: string): Function {
   const initializer = () => {
     const app = injector.get<ApplicationRef>(ApplicationRef);
-    if (!('serviceWorker' in navigator)) {
+    if (!isPlatformBrowser(platformId) || !('serviceWorker' in navigator)) {
```
Comment (CaerusKaru) [inline packages/service-worker/src/module.ts]: One nit. Can be consolidated into `if (!(isPlatformBrowser(platformId) && ('serviceWorker' in navigator)) {` 

(i.e. "if not browser or service worker not supported" -> "if there is no service worker")

**Comment by Toxicable** [comment]
The travis failure is due to payload-size.js
>Commit 24cf8b326918e0694dfcdcf48635b5325060827a gzip7 service-worker increase from 2540 to 2592
Commit 24cf8b326918e0694dfcdcf48635b5325060827a gzip9 service-worker increase from 2540 to 2592
Commit 24cf8b326918e0694dfcdcf48635b5325060827a uncompressed service-worker increase from 8974 to 9136

Im not sure what to change to let the size increase be valid

**Comment by CaerusKaru** [comment]
Your problem is because [this script](https://github.com/angular/angular/blob/f7c9b941cbfe8e9ff127977145a5b75538622093/scripts/ci/payload-size.js) reads a greater than 1% increase and errors out.

**Comment by CaerusKaru** [comment]
You can update the script to allow for a higher variance, but I'm not sure that's the intention

**Comment by Toxicable** [comment]
Yes I know that's the cause, I pointed that out in my previous comment.

**Comment by CaerusKaru** [comment]
I updated my comment to reflect that, I didn't realize you linked to the responsible script. Apologies.

**Comment by CaerusKaru** [comment]
One method to get this to pass is to remove the dep on angular/common and just add a simple `navigator` check like I mentioned in the original issue. But this obviously isn't the best solution long-term.

**Comment by jasonaden** [comment]
Yes, you're right (for now) on the 1% size diff. If that's the only failure here, this should get reviewed and could get merged.

**Comment by Toxicable** [comment]
I think that Alex should take over this one.
I've tested by fix locally but to merge into core it should really have a integration test to go with it, which I have no idea how to setup

**Review by alxhub — APPROVED** [review]
Thanks!

**Comment by Toxicable** [comment]
@alxhub Rebased and good to go now

**Comment by Toxicable** [comment]
closed via https://github.com/angular/angular/commit/7cabaa0ae745dfbbad664b71da70eec89c17c744

**Pull request closed without merging**